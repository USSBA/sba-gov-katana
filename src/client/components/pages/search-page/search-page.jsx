import { isEmpty } from 'lodash'
import React, { PureComponent } from 'react'
import { browserHistory } from 'react-router'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { BasicLink, SmallPrimaryButton, TextInput } from 'atoms'
import { Paginator } from 'molecules'
import styles from './search-page.scss'
import { logPageEvent } from '../../../services/analytics.js'
import * as ContentActions from '../../../actions/content.js'

const getQueryParams = search => {
  const decoded = decodeURIComponent(search)
  const formatted = decoded.replace(/\+/g, ' ')

  let searchTerm = formatted.split('q=')[1]
  if (!isEmpty(searchTerm)) {
    searchTerm = searchTerm.indexOf('&') !== -1 ? searchTerm.split('&')[0] : searchTerm
  } else {
    searchTerm = ''
  }

  let pageNumber = formatted.split('p=')[1]
  if (!isEmpty(pageNumber)) {
    pageNumber = Number(pageNumber.indexOf('&') !== -1 ? pageNumber.split('&')[0] : pageNumber)
    pageNumber = isNaN(pageNumber) ? 1 : pageNumber
  } else {
    pageNumber = 1
  }

  return {
    searchTerm,
    pageNumber
  }
}

class SearchPage extends PureComponent {
  constructor() {
    super()

    this.state = {
      searchTerm: '',
      newSearchTerm: '',
      searchResults: [],
      pageNumber: 1,
      pageSize: 10,
      itemCount: 0,
      start: 0
    }
  }

  componentWillMount() {
    // if there is are parameters
    if (!isEmpty(document.location.search)) {
      // on componentMount, pull search term from url
      const { searchTerm, pageNumber } = getQueryParams(document.location.search)

      // also, match the new search term to this current search term
      // this enables the input field and the submit button to function
      // as expected when the browser refreshes the page
      const newSearchTerm = searchTerm

      this.setState(
        {
          searchTerm,
          newSearchTerm,
          pageNumber
        },
        () => {
          let newStartValue = 0
          pageNumber > 1 ? newStartValue = (pageNumber - 1) * 10 : newStartValue = 0

          this.props.actions.fetchContentIfNeeded('search', 'search', {
            term: searchTerm,
            pageNumber: pageNumber,
            pageSize: this.state.pageSize,
            start: newStartValue
          })
        }
      )
    }
  }

  componentWillReceiveProps(nextProps) {
    // The SearchBar child component's browserHistory.push() call
    // triggers this componentWillReceiveProps() lifecyle method

    // on componentWillReceiveProps
    // get the search term from the React Router url
    const { searchTerm } = getQueryParams(nextProps.location.search)

    // update search results
    const { searchResults, itemCount } = nextProps

    this.setState(
      {
        searchTerm,
        searchResults,
        itemCount
      },
      () => {
        if (!isEmpty(nextProps.location.search)) {
          // after the search term state has been updated
          // show it's value in the input field
          // --
          // this enables the input field to update itself
          // when a browser history is navigated
          this.updateSearchInputValue(searchTerm)
        } else {
          this.updateSearchInputValue('')
        }
      }
    )
  }

  updateSearchInputValue(value) {
    const input = document.getElementById('search')
    input.value = value
  }

  onSearchInputChange(newSearchTerm) {
    const encoded = encodeURIComponent(newSearchTerm)
    this.setState({ newSearchTerm: encoded })
  }

  onPageNumberChange(pageNumber) {
    if (pageNumber > 1) {
      this.setState({
        start: (pageNumber - 1) * 10
      })
    } else {
      this.setState({
        start: 0
      })
    }
    this.setState({
      pageNumber
    }, () => {
      this.onSubmit(true)
    })
  }

  onSubmit(resetPageNumber) {
    const { newSearchTerm: term, pageNumber, pageSize, start } = this.state

    if (resetPageNumber === 1) {
      this.setState({
        start: 0
      })
    }

    const data = {
      term,
      pageNumber,
      pageSize,
      start
    }

    this.props.actions.fetchContentIfNeeded('search', 'search', data)

    // browserHistory.push() triggers the HOC componentWillReceiveProps() lifecyle method

    browserHistory.push({
      pathname: '/search',
      search: '?q=' + term + '&p=' + pageNumber
    })
  }

  render() {
    const { searchTerm, newSearchTerm, searchResults } = this.state

    const { hasNoResults } = this.props

    return (
      <div className={styles.container}>
        <h1>Search</h1>
        <SearchBar
          searchTerm={searchTerm}
          newSearchTerm={newSearchTerm}
          onSearchInputChange={this.onSearchInputChange.bind(this)}
          onSubmit={this.onSubmit.bind(this)}
        />
        {!isEmpty(searchTerm) && (
          <div>
            <hr />
            <div>
              {searchResults.length > 0 && (
                <div>
                  <ResultsList {...this.state} onPageNumberChange={this.onPageNumberChange.bind(this)} />
                </div>
              )}
              {!hasNoResults &&
                searchResults.length === 0 && (
                  <div>
                    <p className="results-message">loading...</p>
                  </div>
                )}
              {hasNoResults && (
                <div>
                  <p className="results-message">Sorry, we couldn't find anything matching that query.</p>
                </div>
              )}
            </div>
          </div>
        )}
      </div>
    )
  }
}

const SearchBar = props => {
  const { searchTerm, newSearchTerm, onSearchInputChange, onSubmit } = props

  const submit = term => {
    props.actions.fetchContentIfNeeded('search', 'search', {
      term
    })

    // browserHistory.push() triggers the HOC componentWillReceiveProps() lifecyle method

    browserHistory.push({
      pathname: '/search',
      search: '?q=' + term
    })
  }

  return (
    <div>
      <div className={styles.textInput}>
        <TextInput
          id="search"
          errorText={'Please enter the correct thing.'}
          label="What are you looking for?"
          validationState={''}
          defaultValue={searchTerm}
          onChange={obj => {
            onSearchInputChange(obj.target.value)
          }}
          onKeyDown={obj => {
            const enterKeyCode = 13
            if (obj.keyCode === enterKeyCode && searchTerm !== decodeURIComponent(newSearchTerm)) {
              onSubmit(1)
            }
          }}
          aria-controls="results-list"
        />
      </div>
      <div className={styles.searchButton}>
        <SmallPrimaryButton
          id="submit-button"
          text="Search"
          onClick={() => {
            if (!isEmpty(searchTerm) && searchTerm !== decodeURIComponent(newSearchTerm)) {
              onSubmit(1)
            }
          }}
          aria-controls="results-list"
        />
      </div>
    </div>
  )
}

const ResultsList = props => {
  const { searchTerm, pageNumber, pageSize, searchResults, itemCount, onPageNumberChange } = props

  const handleBack = () => {
    const newPageNumber = Math.max(1, pageNumber - 1)
    onPageNumberChange(newPageNumber)
    logPageEvent({
      category: 'Show-More-Results',
      action: 'Previous'
    })
  }

  const handleForward = () => {
    const newPageNumber = Math.min(Math.max(1, Math.ceil(itemCount / pageSize)), pageNumber + 1)
    onPageNumberChange(newPageNumber)
    logPageEvent({
      category: 'Show-More-Results',
      action: 'Next'
    })
  }

  const renderPaginator = () => {
    return (
      <div className={styles.paginator}>
        <Paginator
          id={"current-total-result-number"}
          pageNumber={pageNumber}
          pageSize={pageSize}
          total={itemCount}
          onBack={() => {
            handleBack()
          }}
          onForward={() => {
            handleForward()
          }}
        />
      </div>
    )
  }

  const renderList = () => {
    return (
      <div className={styles.results}>
        {searchResults.map((item, index) => {
          let title
          let summary
          let url

          if (!isEmpty(item.fields)) {
            title = item.fields.title
            summary = item.fields.summary
            url = item.fields.url
          }

          // only show results with URL
          // if (url) {
            return (
              <div key={index} className={`${styles.result}  result-box`}>
                <div className={styles.title}>
                  <BasicLink url={url} myClassName={"result-title"}>{title}</BasicLink>
                </div>
                <div className={`${styles.summary} result-summary`}>{summary}</div>
                <div className={styles.url}>
                  <BasicLink url={url} myClassName={"result-url"}>{url}</BasicLink>
                </div>
              </div>
            )
          // }
        })}
      </div>
    )
  }

  return (
    <div>
      <div>
        <div className={styles.searchTerm}>
          <span id="search-term-title">"{searchTerm}"</span>
        </div>
        {renderPaginator()}
      </div>
      <div role="region" id="results-list" aria-live="polite" aria-relevant="additions removals">
        {renderList()}
      </div>
      {renderPaginator()}
    </div>
  )
}

function mapReduxStateToProps(reduxState, ownProps) {
  let searchResults = []
  let itemCount = 0
  let hasNoResults = false

  if (!isEmpty(reduxState.contentReducer.search)) {
    searchResults = reduxState.contentReducer.search.hits.hit
    itemCount = reduxState.contentReducer.search.hits.found
    hasNoResults = reduxState.contentReducer.search.hasNoResults
  }

  return {
    searchResults,
    itemCount,
    hasNoResults
  }
}

=======
}

const ResultsList = props => {
  const { searchTerm, pageNumber, pageSize, searchResults, itemCount, onPageNumberChange } = props

  const handleBack = () => {
    const newPageNumber = Math.max(1, pageNumber - 1)
    onPageNumberChange(newPageNumber)
    logPageEvent({
      category: 'Show-More-Results',
      action: 'Previous'
    })
  }

  const handleForward = () => {
    const newPageNumber = Math.min(Math.max(1, Math.ceil(itemCount / pageSize)), pageNumber + 1)
    onPageNumberChange(newPageNumber)
    logPageEvent({
      category: 'Show-More-Results',
      action: 'Next'
    })
  }

  const renderPaginator = () => {
    return (
      <div className={styles.paginator}>
        <Paginator
          pageNumber={pageNumber}
          pageSize={pageSize}
          total={itemCount}
          onBack={() => {
            handleBack()
          }}
          onForward={() => {
            handleForward()
          }}
        />
      </div>
    )
  }

  const renderList = () => {
    return (
      <div className={styles.results}>
        {searchResults.map((item, index) => {
          let title
          let summary
          let url

          if (!isEmpty(item.fields)) {
            title = item.fields.title
            summary = item.fields.summary
            url = item.fields.url
          }

          return (
            <div key={index} className={styles.result}>
              <div className={styles.title}>
                <BasicLink url={url}>{title}</BasicLink>
              </div>
              <div className={styles.summary}>{summary}</div>
              <div className={styles.url}>
                <BasicLink url={url}>{url}</BasicLink>
              </div>
            </div>
          )
        })}
      </div>
    )
  }

  return (
    <div>
      <div>
        <div className={styles.searchTerm}>
          <span>"{searchTerm}"</span>
        </div>
        {renderPaginator()}
      </div>
      <div role="region" id="results-list" aria-live="polite" aria-relevant="additions removals">
        {renderList()}
      </div>
      {renderPaginator()}
    </div>
  )
}

function mapReduxStateToProps(reduxState, ownProps) {
  let searchResults = []
  let itemCount = 0
  let hasNoResults = false

  if (!isEmpty(reduxState.contentReducer.search)) {
    searchResults = reduxState.contentReducer.search.hits.hit
    itemCount = reduxState.contentReducer.search.hits.found
    hasNoResults = reduxState.contentReducer.search.hasNoResults
  }

  return {
    searchResults,
    itemCount,
    hasNoResults
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(ContentActions, dispatch)
  }
}

export default connect(mapReduxStateToProps, mapDispatchToProps)(SearchPage)

export { SearchPage, SearchBar, ResultsList }
