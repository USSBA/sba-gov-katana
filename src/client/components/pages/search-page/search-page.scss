.container {
  @include grid-media($sba-grid-small) {
    @include grid-column(6);
    margin-top: 20px;
    margin-bottom: 40px;
  }
  @include grid-media($sba-grid-medium) {
    @include grid-column(12);
    margin-top: 20px;
    margin-bottom: 30px;
  }
  @include grid-media($sba-grid-large) {
    @include grid-column(8);
    @include grid-push(2);
  }
}

.textInput {
  width: 60%;
  display: inline-block;

  input {
    height: 53px;
  }
}

.searchButton {
  display: inline-block;

  button {
    margin-top: -4px;
    font-size: 22px;
    line-height: 0.75;
  }
}

.searchTerm {
  float: left;
  margin-top: 15px;

  span {
    font-weight: bold;
    font-size: 120%;
  }
}

.paginator {
  float: right;
  margin: 30px 0;
}

.results {
  clear: both;
}

.result {
  border: 1px solid #bebdbd;
  padding: 20px;
  margin-top: 30px;
}

.url a {
  $action-color: #8a8787;
  color: $action-color;

  &:hover {
    color: shade($action-color, 25%);
  }
}
