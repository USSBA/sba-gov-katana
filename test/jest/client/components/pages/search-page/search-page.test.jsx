import React from 'react'
import { SearchPage, SearchBar, ResultsList } from 'client/components/pages/search-page/search-page.jsx'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
// Quiet warnings about OnTouchTap
import injectTapEventPlugin from 'react-tap-event-plugin'
injectTapEventPlugin()

const mockFunction = jest.fn()
const props = {
  actions: {
    fetchContentIfNeeded: mockFunction
  }
}

describe('SearchPage', () => {
  test('to match snapshot', () => {
    const component = renderer.create(<SearchPage {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('SearchBar', () => {
    describe('TextInput', () => {
      it('has one input field', () => {
        const component = mount(<SearchBar />)
        const result = component.find('TextInput').length
        const expected = 1

        expect(result).toEqual(expected)
      })
    })

    describe('Button', () => {
      it('has one submit button', () => {
        const component = mount(<SearchBar />)
        const result = component.find('Button').length
        const expected = 1

        expect(result).toEqual(expected)
      })

      it('the submit button has a label of "Search"', () => {
        const component = mount(<SearchBar />)
        const result = component.find('Button').text()
        const expected = 'Search'

        expect(result).toEqual(expected)
      })
    })
  })

  describe('ResultsList', () => {
    describe('Paginator', () => {
      it('has two paginator instances', () => {
        const props = {
          searchResults: [
            {
              title: 'Title',
              description: 'Description',
              url: '#'
            }
          ],
          pageNumber: 1,
          pageSize: 5,
          itemCount: 1
        }
        const component = mount(<ResultsList {...props} />)
        const result = component.find('Paginator').length
        const expected = 1

        expect(result).toEqual(expected)
      })
    })
  })

  describe('SuggestedRoute', () => {
    it('should display SuggestedRouteCard when there is a search term', () => {
      const component = shallow(<SearchPage {...props} />)
      component.setState({
        searchTerm: 'my search term'
      })
      const result = component.find('SuggestedRouteCard').length
      const expected = 1
      expect(result).toEqual(expected)
    })
    it('should not display SuggestedRouteCard when there is no search term', () => {
      const component = shallow(<SearchPage {...props} />)
      component.setState({
        searchTerm: ''
      })
      const result = component.find('SuggestedRouteCard').length
      const expected = 0
      expect(result).toEqual(expected)
    })
  })
})
