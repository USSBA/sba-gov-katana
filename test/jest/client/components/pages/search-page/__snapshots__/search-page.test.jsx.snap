// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage to match snapshot 1`] = `
<div
  className="container"
>
  <h1>
    Search
  </h1>
  <div>
    <div
      className="textInput"
    >
      <div
        className="inputContainer "
        hidden={undefined}
        id="search-container"
      >
        <label
          className="controlLabel"
          htmlFor="search"
        >
          What are you looking for?
        </label>
        <div
          className="textInputContainer"
        >
          <input
            aria-controls="results-list"
            className="textInput "
            defaultValue=""
            id="search"
            onChange={[Function]}
            onKeyDown={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="searchButton"
    >
      <button
        aria-controls="results-list"
        className="button alternate large primary responsive spacing"
        id="submit-button"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
  </div>
</div>
`;
